lib/writer.ml

```diff
@@ -154,58 +181,152 @@ let assemble_alpn_protocol p =
-  let pay, typ = match e with
-    | `EllipticCurves cs -> (assemble_elliptic_curves cs, ELLIPTIC_CURVES)
+  assemble_ext @@ match e with
+    | `SupportedGroups groups ->
+      (assemble_supported_groups groups, SUPPORTED_GROUPS)
     | `Hostname name -> (assemble_hostnames [name], SERVER_NAME)
     | `Padding x ->
-       let buf = create x in
-       memset buf 0 ;
-       (buf, PADDING)
+      let buf = create x in
+      memset buf 0 ;
+      (buf, PADDING)
     | `SignatureAlgorithms s ->
       (assemble_signature_algorithms s, SIGNATURE_ALGORITHMS)
     | `ALPN protocols ->
       (assemble_alpn_protocols protocols, APPLICATION_LAYER_PROTOCOL_NEGOTIATION)
+    | `KeyShare ks ->
+      (assemble_list Two assemble_keyshare_entry ks, KEY_SHARE)
+    | `PreSharedKeys ids ->
+      (assemble_client_psks ids, PRE_SHARED_KEY)
+    | `EarlyDataIndication ->
+      (create 0, EARLY_DATA)
+    | `SupportedVersions vs ->
+      (assemble_supported_versions vs, SUPPORTED_VERSIONS)
+    | `PostHandshakeAuthentication ->
+      (Utils.Cs.empty, POST_HANDSHAKE_AUTH)
+    | `Cookie c ->
+      (assemble_cookie c, COOKIE)
+    | `PskKeyExchangeModes modes ->
+      (assemble_psk_key_exchange_modes modes, PSK_KEY_EXCHANGE_MODES)
     | x -> assemble_extension x
-  in
-  let buf = create 4 in
-  BE.set_uint16 buf 0 (extension_type_to_int typ);
-  BE.set_uint16 buf 2 (Cstruct.len pay);
-  buf <+> pay
 
 let assemble_server_extension e =
-  let pay, typ = match e with
+  assemble_ext @@ match e with
     | `Hostname -> (create 0, SERVER_NAME)
     | `ALPN protocol ->
       (assemble_alpn_protocols [protocol], APPLICATION_LAYER_PROTOCOL_NEGOTIATION)
+    | `KeyShare (g, ks) ->
+      let ng = group_to_named_group g in
+      (assemble_keyshare_entry (ng, ks), KEY_SHARE)
+    | `PreSharedKey id ->
+      let data = create 2 in
+      BE.set_uint16 data 0 id ;
+      (data, PRE_SHARED_KEY)
+    | `SelectedVersion v -> (assemble_protocol_version v, SUPPORTED_VERSIONS)
     | x -> assemble_extension x
+
+let assemble_encrypted_extension e =
+  assemble_ext @@ match e with
+    | `Hostname -> (create 0, SERVER_NAME)
+    | `ALPN protocol ->
+      (assemble_alpn_protocols [protocol], APPLICATION_LAYER_PROTOCOL_NEGOTIATION)
+    | `SupportedGroups groups ->
+      (assemble_supported_groups (List.map group_to_named_group groups), SUPPORTED_GROUPS)
+    | `EarlyDataIndication -> (create 0, EARLY_DATA)
+    | _ -> invalid_arg "unknown extension"
+
+let assemble_retry_extension e =
+  assemble_ext @@ match e with
+    | `SelectedGroup g -> (assemble_group g, KEY_SHARE)
+    | `Cookie c -> (assemble_cookie c, COOKIE)
+    | `SelectedVersion v -> (assemble_protocol_version v, SUPPORTED_VERSIONS)
+    | `UnknownExtension _ -> invalid_arg "unknown retry extension"
+
+let assemble_cert_ext (certificate, extensions) =
+  let cert = assemble_certificate certificate
+  and exts = assemble_list Two assemble_server_extension (* ?? *) extensions
```

Comment:
What does `(* ?? *)` mean here?

Reply:
when I wrote the code, I did not know whether this was the intended binary representation.